### PR TITLE
Bug 1835004: Prefer metav1.Duration in API types

### PR DESCRIPTION
**Pull request by dmage**
time.Duration is encoded as an integer (nanoseconds), but keepYoungerThan is documented and designed to be a string.

### CODE REVIEW

File: imageregistry/v1/types_imagepruner.go [hunk 1]
```diff
@@ -44,13 +42,13 @@ type ImagePrunerSpec struct {
...
-	// Defaults to 5.
+	// Defaults to 3.
 	// +optional
 	KeepTagRevisions *int `json:"keepTagRevisions,omitempty" protobuf:"bytes,3,opt,name=keepTagRevisions"`
 	// keepYoungerThan specifies the minimum age of an image and its referrers for it to be considered a candidate for pruning.
-	// Defaults to 96h (96 hours).
+	// Defaults to 60m (60 minutes).
 	// +optional
-	KeepYoungerThan *time.Duration `json:"keepYoungerThan,omitempty" protobuf:"varint,4,opt,name=keepYoungerThan,casttype=time.Duration"`
+	KeepYoungerThan *metav1.Duration `json:"keepYoungerThan,omitempty" protobuf:"bytes,4,opt,name=keepYoungerThan"`
```
Comment (adambkaplan) [inline imageregistry/v1/types_imagepruner.go:51]: /cc @sttts @deads2k 

My gut is that we can't do this, as the data is currently serialized as JSON integers.
I think we need to deprecate `KeepYoungerThan` and make a new field - `KeepYoungerThanDuration`

Comment (sttts) [reply imageregistry/v1/types_imagepruner.go:51]: yes, I think that's right.

One could maybe wrap `metav1.Duration` into a local type with a custom unmarshaller that understands the old serialization.

File: imageregistry/v1/01-crd.yaml [hunk 2]
```diff
@@ -619,14 +619,13 @@ spec:
...
+                for a tag in an image stream that will be preserved. Defaults to 3.
               type: integer
             keepYoungerThan:
               description: keepYoungerThan specifies the minimum age of an image and
                 its referrers for it to be considered a candidate for pruning. Defaults
-                to 96h (96 hours).
-              type: integer
-              format: int64
+                to 60m (60 minutes).
+              type: string
```
Comment (sttts) [inline imageregistry/v1/01-crd.yaml:628]: this means that old integers won't validate anymore. There is `x-kubernetes-int-or-string: true`, usually used with the `IntOrString` type, and afaik not supported by controller-tools via annotation. But you can patch it in by using a patch file next to 01-crd.yaml here, compare https://github.com/openshift/api/blob/master/operator/v1/0000_20_kube-apiserver-operator_01_config.crd.yaml-merge-patch.

File: imageregistry/v1/types_imagepruner.go (hunk 1 above)
Comment (deads2k) [reply imageregistry/v1/types_imagepruner.go:51]: > yes, I think that's right.
> 
> One could maybe wrap `metav1.Duration` into a local type with a custom unmarshaller that understands the old serialization.

That only resolves half the problem.  Any other consumer won't understand the resulting serialization if we change it.

Comment (deads2k) [reply imageregistry/v1/types_imagepruner.go:51]: > /cc @sttts @deads2k
> 
> My gut is that we can't do this, as the data is currently serialized as JSON integers.
> I think we need to deprecate `KeepYoungerThan` and make a new field - `KeepYoungerThanDuration`

Can you try to come up with a compatibility plan for this new field that you would add?  I think it needs to consider...

1. old consumer which removes the new field and modifies the old
2. old consumer which keeps the new field and modifies the old
3. new consumer which keeps the old field and modifies the new
4. case where new object is not written (so admission isn't involved) and new consumer attempts a read

Comment (smarterclayton) [reply imageregistry/v1/types_imagepruner.go:51]: Is this set with a default value by users today?  Or would this be empty on a fresh install?

Comment (dmage) [reply imageregistry/v1/types_imagepruner.go:51]: It's empty on a fresh install.

File: imageregistry/v1/01-crd.yaml (hunk 2 above)
Comment (dmage) [reply imageregistry/v1/01-crd.yaml:628]: This fields is ment to be populated by user, and our documentation and godoc give examples like "96h" or "60m". The meaning of this field as an integer is not defined.

**Comment by dmage** [comment]
Replaced by #649.

**Pull request closed without merging**